mbq/client/contrib/permissions.py

```diff
@@ -0,0 +1,149 @@
+        self.os_core_client = os_core_client
+
+        if cache_name is not None:
+            from django.core.cache import caches
+
+            self.cache = caches[cache_name] if cache_name else None
+            self.cache_period_seconds = cache_period_seconds
+        else:
+            self.cache = None
+            self.cache_period_seconds = None
+
+    def _cache_key(
+        self, person_id: str, ref: Union[str, int], ref_type: Optional[RefType] = None
+    ) -> str:
+        """ref can be either an org_ref or a legacy int ref here"""
+        if ref_type:
+            return f"{self._cache_prefix}:{person_id}:{ref}:{ref_type}"
+        return f"{self._cache_prefix}:{person_id}:{ref}"
+
+    def _global_cache_key(self, person_id: str) -> str:
+        return f"{self._cache_prefix}:{person_id}:global"
+
+    def _cache_read(
+        self, person_id: str, org_ref: str, ref_type: Optional[RefType] = None
+    ) -> Optional[CachedPermissionsDoc]:
+        if not self.cache:
+            return None
+
+        keys = [self._global_cache_key(person_id)]
+        if org_ref != "global":
+            keys.append(self._cache_key(person_id, org_ref, ref_type))
+
+        fetched = self.cache.get_many(keys)
+        if len(fetched.keys()) != len(keys):
+            logging.info(f"Not all keys found in cache, got: {fetched}")
+            return None
+
+        logging.info(f"Successful cache read: {fetched}")
+
+        return fetched
+
+    def _cache_write(
+        self, person_id: str, permissions_doc: FetchedPermissionsDoc
+    ) -> CachedPermissionsDoc:
+        if not self.cache:
+            return
+
+        logging.info(f"Transforming to cache representation: {permissions_doc}")
+
+        cache_write_doc = {}
+        for ref, scopes in permissions_doc.items():
+            org_ref, ref_type = ref.split(":")[0], None
+            if ":" in ref:
+                ref_type = ref.split(":")[1]
+            joined_scopes = f"{'|'.join(scopes)}|"
+            cache_write_doc[self._cache_key(person_id, org_ref, ref_type)] = joined_scopes
+
+        logging.info(f"Writing to cache: {cache_write_doc}")
+
+        self.cache.set_many(cache_write_doc, timeout=self.cache_period_seconds)
+
+        return cache_write_doc
+
+    def has_permission(
+        self,
+        person_id: UUIDType,
+        scope: str,
+        org_ref: Union[UUIDType, Literal["global"]],
+        ref_type: Optional[RefType] = None,
+    ) -> bool:
+        person_id = str(person_id)
+        org_ref = str(org_ref)
+
+        cached_doc = self._cache_read(person_id, org_ref, ref_type)
+        if not cached_doc:
+            fetched_doc = self.os_core_client.fetch_permissions(person_id, org_ref, ref_type)
+            cached_doc = self._cache_write(person_id, fetched_doc)
+
+        if cached_doc:
+            for scopes in cached_doc.values():
+                if f"{scope}|" in scopes:
```

Comment:
why include the pipe char here?

Reply:
This protects against prefix collisions. Otherwise, you might end up with something like:

```
if "read:order" in "read:invoice|read:order_details":
    return "Whoops"
```